zcash_client_backend/src/zip321.rs

```diff
@@ -0,0 +1,1046 @@
+                query_params.push(render::str_param(&name, &value, payment_index))
+            }
+        }
+
+        let mut query_params = vec![];
+        if self.payments.len() == 1 {
+            self.payments.get(0).map(|payment| {
+                push_params(payment, None, &mut query_params);
+                format!(
+                    "zcash:{}?{}",
+                    payment.recipient_address.render(params),
+                    query_params.join("&")
+                )
+            })
+        } else {
+            for (i, payment) in (&self.payments).into_iter().enumerate() {
+                query_params.push(render::addr_param(
+                    params,
+                    &payment.recipient_address,
+                    Some(i),
+                ));
+                push_params(payment, Some(i), &mut query_params);
+            }
+
+            Some(format!("zcash:?{}", query_params.join("&")))
+        }
+    }
+
+    /// Parse the provided URI to a payment request value.
+    pub fn from_uri<P: consensus::Parameters>(params: &P, uri: &str) -> Result<Self, String> {
+        // Parse the leading zcash:<address>
+        let (rest, primary_addr_param) =
+            parse::lead_addr(&params.clone())(uri).map_err(|e| e.to_string())?;
+
+        // Parse the remaining parameters as an undifferentiated list
+        let (_, xs) = all_consuming(preceded(
+            char('?'),
+            separated_list(char('&'), parse::zcashparam(params)),
+        ))(rest)
+        .map_err(|e| e.to_string())?;
+
+        // Construct sets of payment parameters, keyed by the payment index.
+        let mut params_by_index: HashMap<usize, Vec<parse::Param>> = HashMap::new();
+
+        // Add the primary address, if any, to the index.
+        for p in primary_addr_param {
+            params_by_index.insert(p.payment_index, vec![p.param]);
+        }
+
+        // Group the remaining parameters by payment index
+        for p in xs {
+            match params_by_index.get_mut(&p.payment_index) {
+                None => {
+                    params_by_index.insert(p.payment_index, vec![p.param]);
+                }
+
+                Some(current) => {
+                    if parse::has_duplicate_param(&current, &p.param) {
+                        return Err(format!(
+                            "Found duplicate parameter {:?} at index {}",
+                            p.param, p.payment_index
+                        ));
+                    } else {
+                        current.push(p.param);
+                    }
+                }
+            }
+        }
+
+        // Build the actual payment values from the index.
+        let mut payments = vec![];
+        for (i, params) in params_by_index {
+            let payment = parse::to_payment(params, i)?;
+            payments.push(payment);
+        }
+
+        Ok(Zip321Request { payments })
+    }
+}
+
+pub mod render {
```

Comment:
I don't see why this needs to be exposed beyond its use in `Zip321Request::to_uri`
```suggestion
mod render {
```